File: libraries/typescript/src/writer.ts
import { IfcxDocument } from './document.js';

/**
 * Writes IFCX (JSON) files
 */
export class IfcxWriter {
  /** Write to formatted JSON string */
  static toString(doc: IfcxDocument, indent = 2): string {
    return JSON.stringify(doc.toJSON(), null, indent);
  }

  /** Write to buffer (UTF-8 encoded JSON) */
  static toBuffer(doc: IfcxDocument, indent = 2): Uint8Array {
    const json = IfcxWriter.toString(doc, indent);
    const encoder = new TextEncoder();
    return encoder.encode(json);
  }

  /** Write compact JSON (no whitespace) */
  static toCompactString(doc: IfcxDocument): string {
    return JSON.stringify(doc.toJSON());
  }
}

File: libraries/rust/src/writer.rs
//! IFCX file writer.

use crate::{IfcxDocument, IfcxError};
use std::path::Path;

pub struct IfcxWriter;

impl IfcxWriter {
    /// Write to formatted JSON string.
    pub fn to_string(doc: &IfcxDocument) -> Result<String, IfcxError> {
        Ok(serde_json::to_string_pretty(&doc.file)?)
    }

    /// Write compact JSON string.
    pub fn to_compact_string(doc: &IfcxDocument) -> Result<String, IfcxError> {
        Ok(serde_json::to_string(&doc.file)?)
    }

    /// Write to file.
    pub fn to_file(doc: &IfcxDocument, path: impl AsRef<Path>) -> Result<(), IfcxError> {
        let json = Self::to_string(doc)?;
        std::fs::write(path, json)?;
        Ok(())
    }
}
